src/lib/shadow-shim-helper-rs/src/rootedcell/rc.rs

```diff
@@ -202,15 +204,32 @@ impl<T> RootedRc<T> {
         }
     }
 
-    /// Safely drop this object, dropping the internal value if no other
-    /// references to it remain.
-    ///
-    /// Instances that are dropped *without* calling this method cannot be
-    /// safely cleaned up. In debug builds this will result in a `panic`.
-    /// Otherwise the underlying reference count will simply not be decremented,
-    /// ultimately resulting in the enclosed value never being dropped.
+    /// Drop the `RootedRc`, and return the inner value if this was the last
+    /// strong reference.
     #[inline]
-    pub fn safely_drop(self, root: &Root) {
+    pub fn into_inner(this: Self, root: &Root) -> Option<T> {
+        this.common.safely_drop(root, RefType::Strong)
+    }
+
+    /// Drops `self`, and if `self` was the last strong reference, call
+    /// `ExplicitDrop::explicit_drop` on the internal value.
+    pub fn explicit_drop_recursive<I, O>(self, root: &Root, param: &I) -> Option<O>
+    where
+        T: ExplicitDrop<ExplicitDropParam = I, ExplicitDropResult = O>,
+    {
+        Self::into_inner(self, root).map(|val| val.explicit_drop(param))
+    }
+}
```

Comment:
Does this need the `I` and `O` generics? I think you could just use `<T as ExplicitDrop>::ExplicitDropParam` and `<T as ExplicitDrop>::ExplicitDropResult` wherever `I` and `O` are used to simplify the type inference?

Reply:
Yup! Good catch!